torch/utils/weak.py

```diff
@@ -1,79 +1,263 @@
+        #   print(w1 == w2)
+        #
+        # This should be False, as a1 and a2 are unrelated (and a1 is
+        # dead anyway)
+        a = self()
+        b = other()
+        if a is not None and b is not None:
+            return a is b
+        return self is other
 
-class WeakTensorKeyDictionary(MutableMapping):
-    data: Dict[WeakTensorRefKey, object]
+# This is directly adapted from cpython/Lib/weakref.py
+class WeakIdKeyDictionary(MutableMapping):
+    data: Dict[WeakIdRef, object]
 
-    def __init__(self):
+    def __init__(self, dict=None):
         self.data = {}
 
-    def __contains__(self, k):
-        return WeakTensorRefKey(k) in self.data
+        def remove(k, selfref=ref(self)):
+            self = selfref()
+            if self is not None:
+                if self._iterating:
+                    self._pending_removals.append(k)
+                else:
+                    try:
+                        del self.data[k]
+                    except KeyError:
+                        pass
+        self._remove = remove
+        # A list of dead weakrefs (keys to be removed)
+        self._pending_removals = []
+        self._iterating = set()
+        self._dirty_len = False
+        if dict is not None:
+            self.update(dict)
 
-    def __len__(self):
-        return len(self.data)
-
-    def __iter__(self):
-        def generator():
-            for wk in self.data:
-                k = wk.ten()
-                if k is not None:
-                    yield k
-        return generator()
-
-    def __getitem__(self, k):
-        return self.data[WeakTensorRefKey(k)]
-
-    def __setitem__(self, k, v):
-        wk = WeakTensorRefKey(k)
-        self_weak_ref = weakref.ref(self)
-
-        def del_ten():
-            self_ref = self_weak_ref()
-            if self_ref is None:
+    def _commit_removals(self):
+        # NOTE: We don't need to call this method before mutating the dict,
+        # because a dead weakref never compares equal to a live weakref,
+        # even if they happened to refer to equal objects.
+        # However, it means keys may already have been removed.
+        pop = self._pending_removals.pop
+        d = self.data
+        while True:
+            try:
+                key = pop()
+            except IndexError:
                 return
-            self_ref.data.pop(wk, None)
-        weakref.finalize(k, del_ten)
-        self.data[wk] = v
 
-    def __delitem__(self, k):
-        del self.data[WeakTensorRefKey(k)]
+            try:
+                del d[key]
+            except KeyError:
+                pass
+
+    def _scrub_removals(self):
+        d = self.data
+        self._pending_removals = [k for k in self._pending_removals if k in d]
+        self._dirty_len = False
+
+    def __delitem__(self, key):
+        self._dirty_len = True
+        del self.data[WeakIdRef(key)]  # CHANGED
+
+    def __getitem__(self, key):
+        return self.data[WeakIdRef(key)]  # CHANGED
```

Comment:
we need to do `_fix_weakref` here. You're calling it elsewhere but if we're going to expose this as a utility we shouldn't require users to call it

Reply:
I disagree. I never take out a strong reference to the key here, so I don't need to fix weakref

Reply:
You're right. Was thinking that data was a WeakValueDictionary